### PR TITLE
CompositeValidationAttribute

### PR DESCRIPTION
**Pull request by AlexejTimonin**
- Add abstract CompositeValidationAttribute.
- Change DataAnnotationsMetadataProvider.CreateValidationMetadata to
populate ValidatorMetadata with validation attributes from CompositeValidationAttribute.

Addresses #7965

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.DataAnnotations/CompositeValidationAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+
+namespace Microsoft.AspNetCore.Mvc.DataAnnotations
+{
+    /// <summary>
+    /// Abstract class for grouping attributes of type <see cref="ValidationAttribute"/> into
+    /// one <see cref="Attribute"/>
+    /// </summary>
+    public abstract class CompositeValidationAttribute : Attribute
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.DataAnnotations/CompositeValidationAttribute.cs]: Could you rename this to `ValidationProviderAttribute`. We want to steer away from using any name DataAnnotations could add in the future.

Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.DataAnnotations/CompositeValidationAttribute.cs]: cc @Eilon \ @mkArtakMSFT  for name

File: src/Microsoft.AspNetCore.Mvc.DataAnnotations/CompositeValidationAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    {
+        /// <summary>
+        /// Get attributes associated with this attribute.
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.DataAnnotations/CompositeValidationAttribute.cs]: > Gets one or more `<see cref="ValidationAttribute" />` instances associated with this attribute.

File: src/Microsoft.AspNetCore.Mvc.DataAnnotations/CompositeValidationAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+        /// </summary>
+        /// <returns><see cref="IEnumerable{ValidationAttribute}"/> associated with this attribute.</returns>
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.DataAnnotations/CompositeValidationAttribute.cs]: I think `IEnumerable{T}` doesn't actually show up correctly in the docs. You could call say>

> Sequence of `<see cref="ValidationAttribute" />` 

or 

>One or more `<see cref="ValidationAttribute" />` ..

File: src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/DataAnnotationsMetadataProvider.cs [hunk 4]
```diff
@@ -317,15 +317,24 @@ public void CreateValidationMetadata(ValidationMetadataProviderContext context)
                 throw new ArgumentNullException(nameof(context));
             }
 
+            var attributes = context.Attributes.ToList();
+
+            var attributesInComposite = new List<object>();
+            foreach (var compositeAttribute in attributes.OfType<CompositeValidationAttribute>())
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/DataAnnotationsMetadataProvider.cs]:
```C#
var count = attributes.Count;
for (var i= 0; i < count; i++)
{
    if (attributes[i] is ValidationProviderAttribute validationProviderAttribute) { attributes.AddRange(validationProviderAttribute.GetAttributes()); }
}
```

File: src/Microsoft.AspNetCore.Mvc.DataAnnotations/CompositeValidationAttribute.cs [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+        public abstract IEnumerable<ValidationAttribute> GetAttributes();
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.DataAnnotations/CompositeValidationAttribute.cs]: Consider naming this `GetValidationAttributes()`

File: test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataAnnotationsMetadataProviderTest.cs [hunk 6]
```diff
@@ -1241,6 +1241,38 @@ public void CreateValidationMetadata_NoRequiredAttribute_IsRequiredLeftAlone(boo
             Assert.Equal(initialValue, context.ValidationMetadata.IsRequired);
         }
 
+        [Fact]
+        public void CreateValidationMetadata_WillAddValidationAttributes_From_CompositeValidationAttribute()
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataAnnotationsMetadataProviderTest.cs]: These tests look good. However we should add a few functional tests that verify this E2E (including verifying if client side validation) is correctly inferred correctly.

Comment (AlexejTimonin) [reply test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataAnnotationsMetadataProviderTest.cs]: Thank you for your feedback! Can I add the tests in InputObjectValidationTests.cs or should I create new test class for this?

File: src/Microsoft.AspNetCore.Mvc.DataAnnotations/ValidationProviderAttribute.cs [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+namespace Microsoft.AspNetCore.Mvc.DataAnnotations
+{
+    /// <summary>
+    /// Abstract class for grouping attributes of type <see cref="ValidationAttribute"/> into
+    /// one <see cref="Attribute"/>
+    /// </summary>
+    public abstract class ValidationProviderAttribute : Attribute
+    {
+        /// <summary>
+        /// Gets one or more <see cref="ValidationAttribute" /> instances associated with this attribute.
```
Comment (Eilon) [inline src/Microsoft.AspNetCore.Mvc.DataAnnotations/ValidationProviderAttribute.cs]: It's really `zero or more`, isn't it?

I suggest simplifying to say `Gets <see cref="ValidationAttribute" /> instances associated with this attribute.`

And people can figure it out on their own.

Comment (AlexejTimonin) [reply src/Microsoft.AspNetCore.Mvc.DataAnnotations/ValidationProviderAttribute.cs]: Thank you for the review! You're correct, it's zero or more. I'll add your suggestion right away.

**Comment by pranavkm** [comment]
Merged in https://github.com/aspnet/Mvc/commit/d4beab5d09bb28e970063f7da9fe916b99514c5c. Thanks for the PR @AlexejTimonin

**Pull request closed without merging**